### ISSUE TITLE
${String[A String Goes here].Compare[Another String Here]} unexpected results. 

### ISSUE THREAD

**Issue by jessebevil** [macroquest/macroquest2#2]
You may or may not have seen the post on MQ2 Forum where I posted this bug. 

http://www.macroquest2.com/phpBB3/viewtopic.php?f=28&t=20331

I reference unexpected results on that thread, and show an output of a loop of comparisons. Along with the offending Sub. 

http://macroquest2.com/wiki/index.php/DataType:string

shows .Compare[] as returning -1, 0, or 1 respective to the comparison results. However, I've narrowed down additional information and feel this is right place to post it. If /echo ${String[Z].Compare[A]} returns 26. for example. But /echo ${String[Zaaaaaa].Compare[Azzzzzz]} then returns a negative value. Even through Zaaaaaa should be after azzzzzzz for example. I replied to my own post on the first link showing my workaround code using Compare.Left[${i}] in a /goto loop until the comparison generates a result that isn't 0. So it only does a comparison until the letters aren't the same, then it does a final check. I don't know C, and I don't know know where the compare method/function/sub or whatever it's called in C is located to recommend a change. (Plus naming conventions for C are also beyond me.) Not sure how long it has been this way. Found it while making a Bank Sorting macro. 

Thanks for the consideration, and the fix if it needs it. 

```
Sub SortBankArray
	/declare i int inner
	/declare j int inner
	/declare k int inner
	/declare temps string inner ""
	/declare tempi string inner ""
	/varset i 1
	/varset j 1
	/varset k 1
	
	

		:iLoop
		:jLoop
		|/echo ${Unsorted[${i},2].Left[1]} compared to ${Unsorted[${j},2]} is ${Unsorted[${i},2].Compare[${Unsorted[${j},2]}]}
		|/echo ${i} ${j}
	/if (!${Unsorted[${i},2].Compare[${Unsorted[${j},2]}]} == 0 && ${i} != ${j}) {
			:kayloop
			/if (${Unsorted[${i},2].Left[${k}].Compare[${Unsorted[${j},2].Left[${k}]}]} == 0) {
				|/echo The first ${k} letters are the same!
				/if (${k} <= ${Unsorted[${i},2].Length}) {
					/varset k ${Int[${Math.Calc[${k}+1]}]}
					/goto :kayloop
				}
		} else /if (${Unsorted[${i},2].Left[${k}].Compare[${Unsorted[${j},2].Left[${k}]}]} > 0) {
|** 			|/echo ${i} is after ${j}
			/varset temps ${Unsorted[${i},2]}
			/varset tempi ${Unsorted[${i},3]} 
			/varset Unsorted[${i},2] ${Unsorted[${j},2]}  
			/varset Unsorted[${i},3] ${Unsorted[${j},3]}  
			/varset Unsorted[${j},2] ${temps} 
			/varset Unsorted[${j},3] ${tempi} **|
			/nomodkey /shiftkey /itemnotify ${Unsorted[${i},1]} leftmouseup
			/nomodkey /shiftkey /itemnotify ${Unsorted[${j},1]} leftmouseup
			/nomodkey /shiftkey /itemnotify ${Unsorted[${i},1]} leftmouseup
			/if (${j} < ${Occupied}) {
				/varset j ${Int[${Math.Calc[${j}+1]}]}
				/goto :jLoop
			} else /if (${i} <= ${Int[${Math.Calc[${Occupied}-1]}]}) {
				/varset i ${Int[${Math.Calc[${i}+1]}]}
				/varset j 2
				/goto :iLoop
			}
		} else /if (${Unsorted[${i},2].Left[${k}].Compare[${Unsorted[${j},2].Left[${k}]}]} < 0) {
			|/echo ${i} is before ${j}
			/if (${i} <= ${Int[${Math.Calc[${Occupied}-1]}]}) {
				/varset i ${Int[${Math.Calc[${i}+1]}]}
				/varset j 2
				/goto :iLoop
			}
		}
	}
	/if (${PopupTimer} <= 0) {
		/popup  Calculating Sort: ${Int[${Math.Calc[100/${Occupied}*${i}]}]}% Complete. Run Time: ${Macro.RunTime} seconds!
		/varset PopupTimer 1s
	}
	/if (${j} < ${Occupied}) {
		/varset j ${Int[${Math.Calc[${j}+1]}]}
		/goto :jLoop
	} else /if (${i} < ${Int[${Math.Calc[${Occupied}-1]}]}) {
		/varset i ${Int[${Math.Calc[${i}+1]}]}
		/varset j 2
		/goto :iLoop
	}
	/echo Done Sorting Array at ${Macro.RunTime} seconds!


	/for i 1 to ${Occupied}
		/echo ${Unsorted[${i},2]} was at ${Unsorted[${i},1]} should be moved to ${Unsorted[${i},3]} 
	/next i

/return
```

**Comment by mackal** [comment]
Compare is using custom Microsoft API call to _stricmp, which does not dictate the return value being -1, 0, or 1. It appears to actually return the distance between the two.

The wiki was never updated to correctly reflect this.

https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/stricmp-wcsicmp-mbsicmp-stricmp-l-wcsicmp-l-mbsicmp-l

Is the documentation, it can still be used for lexigraphical sorting.

I'm also not having issues and I am getting the results I expect (at least to negative, 0, and positive)

You are also giving examples with non-valid mainline MQ2, which tells me that maybe the source of your issue as well.

**Comment by jessebevil** [comment]
If by non valid mainline you mean I don't have a sub main then I do. I simply posted the sub i was using and not the entire macro. If not, then I'm unsure what you mean by non-valid mainline. I did notice it was showing a distance, but in the thread linked I was returning negative when i expected postives. If you say it is working as intend then I'll take your word on it. I've not really got very deep into sorting in macros. Likely as I progress with coding classes Ill learn c and be able to just work with custom plugins. Thanks for the response.
